# Architecture experiment: Subjective and physiological responses to interiors


### Libraries

- [scikit-learn](http://scikit-learn.org/stable/)
- pandas
- matplotlib
- seaborn
- keras

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import OneHotEncoder
from pandas.plotting import scatter_matrix
from sklearn.model_selection import cross_val_predict
from sklearn.model_selection import cross_val_score
import seaborn as sns
%matplotlib inline

In [2]:
DATA_FOLDER = './data/' #The data is in the .gitignore in order to not upload it to the GitHub repository

Execute the following line to export from the jupyter notebook(.ipynb) to a .py file (ignore the warnings):

In [3]:
#!jupyter nbconvert --to script Project2-Notebook.ipynb 

For not show the warnings:

In [4]:
import warnings
warnings.filterwarnings(action = 'once')

## Exploratory Data Analysis

### Loading phase

First we load the data into a Pandas data frame (with the Pandas library; to install this package with conda run:
`conda install -c anaconda pandas`):

In [5]:
df_interior_features = pd.read_csv(DATA_FOLDER + 'T_features.csv')
df_labels_and_features = pd.read_csv(DATA_FOLDER + 'table_dataset_GreeceSwitzerland_N265_metrics_mSC5_JPEGtoBMP_Michelson_RMS.csv')

Let's see what is inside `table_dataset_GreeceSwitzerland_N265_metrics_mSC5_JPEGtoBMP_Michelson_RMS.csv`:

In [6]:
df_labels_and_features.head(7)

ID Country Stimulus_SkyType Stimulus_Context Gender Pattern  pleasant  \
0   1  Greece        clear low           social   Male    P1EL         8   
1   1  Greece        clear low           social   Male    P2EL         8   
2   1  Greece        clear low           social   Male    P3EL         9   
3   1  Greece        clear low           social   Male    P4EL         7   
4   1  Greece        clear low           social   Male    P5EL         8   
5   1  Greece        clear low           social   Male    P6EL         8   
6   2  Greece        clear low           social   Male    P1EL         6   

   interesting  exciting  calming  complex  bright  view  spacious  \
0            7         6        9        4       6     4         8   
1            5         6        8        4       7     3         8   
2            8         8        9        6       7     2         8   
3            7         7        7        6       6     1         8   
4           10         9        7        8       7     5         8   
5            8         8        7        9       7     3         8   
6            7         8        6        0       7     6         7   

   contrast_mean_mSC5  contrast_mean_Michelson  contrast_mean_RMS  \
0           11.188207                 0.907864         106.363985   
1           11.941067                 0.904661         106.590842   
2           12.509719                 0.905066         106.604335   
3           12.846015                 0.906583         106.929875   
4           12.211881                 0.907546         105.478575   
5           12.288385                 0.912304         106.582042   
6           11.188207                 0.907864         106.363985   

   complexity_mean_JPEGtoBMP  
0                   0.104734  
1                   0.122357  
2                   0.119979  
3                   0.115287  
4                   0.118572  
5                   0.113396  
6                   0.104734

In [7]:
df_labels_and_features.shape

(1590, 18)

In [8]:
df_labels_and_features.columns

Index(['ID', 'Country', 'Stimulus_SkyType', 'Stimulus_Context', 'Gender',
       'Pattern', 'pleasant', 'interesting', 'exciting', 'calming', 'complex',
       'bright', 'view', 'spacious', 'contrast_mean_mSC5',
       'contrast_mean_Michelson', 'contrast_mean_RMS',
       'complexity_mean_JPEGtoBMP'],
      dtype='object')

The data frame has 1590 rows × 18 columns where we can see the original sample of the experiment (both inputs and outputs (xn,yn)). Each column is explained below:
* 'ID': Identifier of the subject of the experiment (of one person).
* 'Country': Country where the person was living.  2 x countries (Greece –138 participants, Switzerland –127 participants)
* 'Stimulus_SkyType': 3 x sky types (clear sky with high sun angle, clear sky with low sun angle, overcast sky)
* 'Stimulus_Context':  2 x context scenarios (social scenario, work scenario)
* 'Gender': Male or Female
* 'Pattern': 6 x patterns (determine how are the blinds, and so how are the shadows)
* Outputs (yn): 'pleasant', 'interesting', 'exciting', 'calming', 'complex', 'bright', 'view', 'spacious'.
* interior metrics (for describing the interior (the kind of room, light, ...) quantitatively): 'contrast_mean_mSC5', 'contrast_mean_Michelson', 'contrast_mean_RMS', 'complexity_mean_JPEGtoBMP'.

Note: The metrics in this data set are applied to the whole virtual reallity image (in every interior).

Now let's see what is inside `T_features.csv`:

In [9]:
df_interior_features.head()

filename Pattern Context    SkyType  \
0  p1_social_clearhigh_sg_largewin_simu    P1EL  social  clearhigh   
1  p1_social_clearlow_sg_largewin_simu     P1EL  social   clearlow   
2  p1_social_overcast_sg_largewin_simu     P1EL  social   overcast   
3  p1_work_clearhigh_sg_largewin_simu      P1EL    work  clearhigh   
4  p1_work_clearlow_sg_largewin_simu       P1EL    work   clearlow   

   mean_pleasant  median_pleasant  mean_interesting  median_interesting  \
0       5.439024                6          4.512195                   4   
1       5.571429                6          4.510204                   5   
2       6.093023                6          4.790698                   5   
3       5.804348                6          5.086957                   5   
4       5.255814                5          4.418605                   5   

   mean_calming  median_calming              ...               \
0      5.731707             6.0              ...                
1      5.897959             6.0              ...                
2      6.651163             7.0              ...                
3      6.152174             6.5              ...                
4      5.837209             6.0              ...                

   contrast_max_mSC5  contrast_mean_Michelson  \
0                 82                 0.881714   
1                 85                 0.907864   
2                 49                 0.886125   
3                116                 0.924177   
4                116                 0.937813   

   contrast_mean_Michelson_cube123  contrast_mean_Michelson_cube3  \
0                         0.822095                       0.696552   
1                         0.846031                       0.750000   
2                         0.829098                       0.713287   
3                         0.874757                       0.680272   
4                         0.887639                       0.728223   

   contrast_mean_RMS  contrast_mean_RMS_cube123  contrast_mean_RMS_cube3  \
0         104.887032                 116.926766               138.288359   
1         106.363985                 108.835770               119.479999   
2         102.615246                 113.078576               133.037937   
3         107.051167                 119.883534               142.784796   
4         108.174380                 112.987399               126.132718   

   complexity_mean_JPEGtoBMP  complexity_cube3_JPEGtoBMP  \
0                   0.101518                    0.158865   
1                   0.104734                    0.136334   
2                   0.093733                    0.134794   
3                   0.099205                    0.159858   
4                   0.102264                    0.139392   

   complexity_cube123_JPEGtoBMP  
0                      0.129797  
1                      0.127709  
2                      0.113091  
3                      0.124465  
4                      0.122615  

[5 rows x 31 columns]

In [10]:
df_interior_features.shape

(36, 31)

In [11]:
df_interior_features.columns

Index(['filename', 'Pattern', 'Context', 'SkyType', 'mean_pleasant',
       'median_pleasant', 'mean_interesting', 'median_interesting',
       'mean_calming', 'median_calming', 'mean_exciting', 'median_exciting',
       'mean_complex', 'median_complex', 'mean_bright', 'median_bright',
       'mean_view', 'median_view', 'mean_spacious', 'median_spacious',
       'contrast_mean_mSC5', 'contrast_max_mSC5', 'contrast_mean_Michelson',
       'contrast_mean_Michelson_cube123', 'contrast_mean_Michelson_cube3',
       'contrast_mean_RMS', 'contrast_mean_RMS_cube123',
       'contrast_mean_RMS_cube3', 'complexity_mean_JPEGtoBMP',
       'complexity_cube3_JPEGtoBMP', 'complexity_cube123_JPEGtoBMP'],
      dtype='object')

We have obtained a data frame with 36 rows × 31 columns. This time we see the features of each kind of interior (the kind of room, light, ...)  and several metrics for describing quantitatively the interior situation Each column is explained below:
* 'filename': File where we have the cube map projections associated to every interior. (See in `./Cubemap_Projections/BMP/<filename>` or `./Cubemap_Projections/JPEG/<filename>`). The filenames have the structure "p(pattern_id)_(context)_(SkyType)_sg_largewin_simu".
* 'Pattern': 6 x patterns (determine how are the blinds, and so how are the shadows)
* 'Context': 2 x context scenarios (social scenario, work scenario)
* 'SkyType': 3 x sky types (clear sky with high sun angle, clear sky with low sun angle, overcast sky)
* Means and medians of the outputs (of what people have answer in their respective survey): 'mean_pleasant', 'median_pleasant', 'mean_interesting', 'median_interesting', 'mean_calming', 'median_calming', mean_exciting', 'median_exciting', 'mean_complex', 'median_complex', 'mean_bright', 'median_bright', 'mean_view', 'median_view', 'mean_spacious', 'median_spacious',
* Interior metrics (for describing the interior quantitatively): 'contrast_mean_mSC5', 'contrast_max_mSC5', 'contrast_mean_Michelson', 'contrast_mean_Michelson_cube123', 'contrast_mean_Michelson_cube3', 'contrast_mean_RMS', 'contrast_mean_RMS_cube123','contrast_mean_RMS_cube3', 'complexity_mean_JPEGtoBMP', 'complexity_cube3_JPEGtoBMP', 'complexity_cube123_JPEGtoBMP'

Note: The metrics in this data set are applied not just to the whole virtual reallity image (we have already this data in the other data frame), but to the different parts of the whole image. Cube1 refers to what you see on your left (when you are doing the experiment with VR), cube2 refers to the front and cube3 refers to the right. Cube123 refers to the metric applied to the three subimages. For more details see`./Cubemap_Projections/JPEG/<filename>`.

Having done this previous load and brief analysis, we are going to create the data frame we are interested in work with. Firstly we find that our inputs (or features) could be classified in two types and we are taking both of them: 
* 1- Features of the people (country and gender) who have taken part in the experiment.
* 2- Features of the interior: Now we are taking both categorical and metrics data.

Secondly, the labels (outputs) of our new data frame are the data that comes from every survey (which has been done by each person): 'pleasant', 'interesting', 'exciting', 'calming', 'complex', 'bright', 'view', 'spacious'. We are asked to study first the 'exciting' and the 'calming' labels but we will split the data frame afterwards.

In [12]:
df_labels_and_features.sample()

ID      Country Stimulus_SkyType Stimulus_Context  Gender Pattern  \
1583  126  Switzerland        clear low           social  Female    P6EL   

      pleasant  interesting  exciting  calming  complex  bright  view  \
1583         0            6         2        0        8       4     1   

      spacious  contrast_mean_mSC5  contrast_mean_Michelson  \
1583        10           12.288385                 0.912304   

      contrast_mean_RMS  complexity_mean_JPEGtoBMP  
1583         106.582042                   0.113396

In [20]:
#df_labels_and_features[df_labels_and_features['ID']==21] #What is the ID meaning?

In [14]:
df_interior_features.head(1)

filename Pattern Context    SkyType  \
0  p1_social_clearhigh_sg_largewin_simu    P1EL  social  clearhigh   

   mean_pleasant  median_pleasant  mean_interesting  median_interesting  \
0       5.439024                6          4.512195                   4   

   mean_calming  median_calming              ...               \
0      5.731707             6.0              ...                

   contrast_max_mSC5  contrast_mean_Michelson  \
0                 82                 0.881714   

   contrast_mean_Michelson_cube123  contrast_mean_Michelson_cube3  \
0                         0.822095                       0.696552   

   contrast_mean_RMS  contrast_mean_RMS_cube123  contrast_mean_RMS_cube3  \
0         104.887032                 116.926766               138.288359   

   complexity_mean_JPEGtoBMP  complexity_cube3_JPEGtoBMP  \
0                   0.101518                    0.158865   

   complexity_cube123_JPEGtoBMP  
0                      0.129797  

[1 rows x 31 columns]

For doing the join, we check that one of the metrics determine totally the interior (there are not two interiors with the same metric value):

In [15]:
df_interior_features["to_join"]=(1000*df_interior_features["contrast_mean_RMS"]).astype(int) #We cannot do the join on the float
df_labels_and_features["to_join"]=(1000*df_labels_and_features["contrast_mean_RMS"]).astype(int) #so we add another column

df_interior_features.set_index("to_join").index.is_unique

True

In [16]:
df_labels_and_features.columns

Index(['ID', 'Country', 'Stimulus_SkyType', 'Stimulus_Context', 'Gender',
       'Pattern', 'pleasant', 'interesting', 'exciting', 'calming', 'complex',
       'bright', 'view', 'spacious', 'contrast_mean_mSC5',
       'contrast_mean_Michelson', 'contrast_mean_RMS',
       'complexity_mean_JPEGtoBMP', 'to_join'],
      dtype='object')

In [17]:
df_ml_raw = df_interior_features.merge(df_labels_and_features, on = "to_join", how = 'inner', suffixes = ("_a",""))
df_ml_raw.head(1)

filename Pattern_a Context    SkyType  \
0  p1_social_clearhigh_sg_largewin_simu      P1EL  social  clearhigh   

   mean_pleasant  median_pleasant  mean_interesting  median_interesting  \
0       5.439024                6          4.512195                   4   

   mean_calming  median_calming            ...              exciting  calming  \
0      5.731707             6.0            ...                     5        7   

   complex  bright  view  spacious  contrast_mean_mSC5  \
0        0       7     6         9            5.023496   

   contrast_mean_Michelson  contrast_mean_RMS  complexity_mean_JPEGtoBMP  
0                 0.881714         104.887032                   0.101518  

[1 rows x 50 columns]

In [18]:
df_ml_raw.columns

Index(['filename', 'Pattern_a', 'Context', 'SkyType', 'mean_pleasant',
       'median_pleasant', 'mean_interesting', 'median_interesting',
       'mean_calming', 'median_calming', 'mean_exciting', 'median_exciting',
       'mean_complex', 'median_complex', 'mean_bright', 'median_bright',
       'mean_view', 'median_view', 'mean_spacious', 'median_spacious',
       'contrast_mean_mSC5_a', 'contrast_max_mSC5',
       'contrast_mean_Michelson_a', 'contrast_mean_Michelson_cube123',
       'contrast_mean_Michelson_cube3', 'contrast_mean_RMS_a',
       'contrast_mean_RMS_cube123', 'contrast_mean_RMS_cube3',
       'complexity_mean_JPEGtoBMP_a', 'complexity_cube3_JPEGtoBMP',
       'complexity_cube123_JPEGtoBMP', 'to_join', 'ID', 'Country',
       'Stimulus_SkyType', 'Stimulus_Context', 'Gender', 'Pattern', 'pleasant',
       'interesting', 'exciting', 'calming', 'complex', 'bright', 'view',
       'spacious', 'contrast_mean_mSC5', 'contrast_mean_Michelson',
       'contrast_mean_RMS', 'compl

# New approach

Now we drop the duplicates columns and order the columns in order to have a data frame with the structure X|Y, where X is the matrix of features (each column is a feature) and Y is the matrix with the labels. 

In this new approach we are not using gender.

In [50]:
df_ml = df_ml_raw[['Country', 'Pattern', 'Context', 'SkyType', 'contrast_mean_mSC5', 'contrast_max_mSC5', 
               'contrast_mean_Michelson', 'contrast_mean_Michelson_cube123', 'contrast_mean_Michelson_cube3', 
               'contrast_mean_RMS', 'contrast_mean_RMS_cube123', 'contrast_mean_RMS_cube3', 'complexity_mean_JPEGtoBMP',
               'complexity_cube3_JPEGtoBMP', 'complexity_cube123_JPEGtoBMP', 'pleasant', 'interesting', 'exciting', 'calming', 
               'complex', 'bright', 'view', 'spacious']]

In [51]:
df_ml.head(1)

Country Pattern Context    SkyType  contrast_mean_mSC5  contrast_max_mSC5  \
0  Greece    P1EL  social  clearhigh            5.023496                 82   

   contrast_mean_Michelson  contrast_mean_Michelson_cube123  \
0                 0.881714                         0.822095   

   contrast_mean_Michelson_cube3  contrast_mean_RMS    ...     \
0                       0.696552         104.887032    ...      

   complexity_cube3_JPEGtoBMP  complexity_cube123_JPEGtoBMP  pleasant  \
0                    0.158865                      0.129797         7   

   interesting  exciting  calming  complex  bright  view  spacious  
0            5         5        7        0       7     6         9  

[1 rows x 23 columns]

We can also check and see that there are no missing values. (all instances are non-null)

In [52]:
df_ml.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1590 entries, 0 to 1589
Data columns (total 23 columns):
Country                            1590 non-null object
Pattern                            1590 non-null object
Context                            1590 non-null object
SkyType                            1590 non-null object
contrast_mean_mSC5                 1590 non-null float64
contrast_max_mSC5                  1590 non-null int64
contrast_mean_Michelson            1590 non-null float64
contrast_mean_Michelson_cube123    1590 non-null float64
contrast_mean_Michelson_cube3      1590 non-null float64
contrast_mean_RMS                  1590 non-null float64
contrast_mean_RMS_cube123          1590 non-null float64
contrast_mean_RMS_cube3            1590 non-null float64
complexity_mean_JPEGtoBMP          1590 non-null float64
complexity_cube3_JPEGtoBMP         1590 non-null float64
complexity_cube123_JPEGtoBMP       1590 non-null float64
pleasant                           1590 non-

and see that all ratings are between 0 and 10:

In [53]:
df_ml[['pleasant', 'interesting', 'exciting', 'calming', 'complex', 'bright',
       'view', 'spacious']].describe().loc[['min','max']]


pleasant  interesting  exciting  calming  complex  bright  view  spacious
min       0.0          0.0       0.0      0.0      0.0     0.0   0.0       0.0
max      10.0         10.0      10.0     10.0     10.0    10.0  10.0      10.0

Finally, to work with the new approach we need to compute the percentage of ratings  of every scene that are over a certain value (6, 7 or 8 for example). First we take a new data frame with just the 36 different scenes (72 if we are considering the country):

In [54]:
#x_fdata = df_ml.iloc[:,1:15] # For not considering the country
#x_fdata = df_ml.iloc[:,0:15] # For considering the country

x_fdata = df_ml.iloc[:,0:15] #fdata -> first data without preprocessing
x_fdata.drop_duplicates(inplace = True)
x_fdata

Country Pattern Context    SkyType  contrast_mean_mSC5  \
0          Greece    P1EL  social  clearhigh            5.023496   
21    Switzerland    P1EL  social  clearhigh            5.023496   
41         Greece    P1EL  social   clearlow           11.188207   
68    Switzerland    P1EL  social   clearlow           11.188207   
90         Greece    P1EL  social   overcast            4.458815   
112   Switzerland    P1EL  social   overcast            4.458815   
133        Greece    P1EL    work  clearhigh            5.362578   
156   Switzerland    P1EL    work  clearhigh            5.362578   
179        Greece    P1EL    work   clearlow           11.038933   
202   Switzerland    P1EL    work   clearlow           11.038933   
222        Greece    P1EL    work   overcast            4.916741   
244   Switzerland    P1EL    work   overcast            4.916741   
265        Greece    P2EL  social  clearhigh            4.773985   
286   Switzerland    P2EL  social  clearhigh            4.773985   
306        Greece    P2EL  social   clearlow           11.941067   
333   Switzerland    P2EL  social   clearlow           11.941067   
355        Greece    P2EL  social   overcast            4.375763   
377   Switzerland    P2EL  social   overcast            4.375763   
398        Greece    P2EL    work  clearhigh            5.132830   
421   Switzerland    P2EL    work  clearhigh            5.132830   
444        Greece    P2EL    work   clearlow           11.541689   
467   Switzerland    P2EL    work   clearlow           11.541689   
487        Greece    P2EL    work   overcast            4.862252   
509   Switzerland    P2EL    work   overcast            4.862252   
530        Greece    P3EL  social  clearhigh            5.790874   
551   Switzerland    P3EL  social  clearhigh            5.790874   
571        Greece    P3EL  social   clearlow           12.509719   
598   Switzerland    P3EL  social   clearlow           12.509719   
620        Greece    P3EL  social   overcast            5.028533   
642   Switzerland    P3EL  social   overcast            5.028533   
...           ...     ...     ...        ...                 ...   
928        Greece    P4EL    work  clearhigh            7.087585   
951   Switzerland    P4EL    work  clearhigh            7.087585   
974        Greece    P4EL    work   clearlow           12.490844   
997   Switzerland    P4EL    work   clearlow           12.490844   
1017       Greece    P4EL    work   overcast            6.326519   
1039  Switzerland    P4EL    work   overcast            6.326519   
1060       Greece    P5EL  social  clearhigh            6.702222   
1081  Switzerland    P5EL  social  clearhigh            6.702222   
1101       Greece    P5EL  social   clearlow           12.211881   
1128  Switzerland    P5EL  social   clearlow           12.211881   
1150       Greece    P5EL  social   overcast            5.819852   
1172  Switzerland    P5EL  social   overcast            5.819852   
1193       Greece    P5EL    work  clearhigh            7.011081   
1216  Switzerland    P5EL    work  clearhigh            7.011081   
1239       Greece    P5EL    work   clearlow           12.184296   
1262  Switzerland    P5EL    work   clearlow           12.184296   
1282       Greece    P5EL    work   overcast            6.309748   
1304  Switzerland    P5EL    work   overcast            6.309748   
1325       Greece    P6EL  social  clearhigh            7.130104   
1346  Switzerland    P6EL  social  clearhigh            7.130104   
1366       Greece    P6EL  social   clearlow           12.288385   
1393  Switzerland    P6EL  social   clearlow           12.288385   
1415       Greece    P6EL  social   overcast            6.093422   
1437  Switzerland    P6EL  social   overcast            6.093422   
1458       Greece    P6EL    work  clearhigh            7.406993   
1481  Switzerland    P6EL    work  clearhigh            7.406993   
1504       Greece    P6EL    work   clearlow           12.183556  

In [55]:
#data.groupby(func, axis=0).mean()
df_ml.set_index(['Country', 'Pattern', 'Context', 'SkyType'], inplace = True)

In [62]:
df_ml.loc[['Greece','P1EL', 'social', 'clearhigh']]

contrast_mean_mSC5  contrast_max_mSC5  \
Country Pattern Context SkyType                                            
Greece  P1EL    social  clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearhigh            5.023496                 82   
                        clearlow            11.188207                 85   
                        clearlow            11.188207                 85   
                        clearlow            11.188207                 85   
                        clearlow            11.188207                 85   
                        clearlow            11.188207                 85   
                        clearlow            11.188207                 85   
                        clearlow            11.188207                 85   
                        clearlow            11.188207                 85   
                        clearlow            11.188207                 85   
...                                               ...                ...   
        P6EL    work    clearlow            12.183556                117   
                        clearlow            12.183556                117   
                        clearlow            12.183556                117   
                        clearlow            12.183556                117   
                        clearlow            12.183556                117   
                        clearlow            12.183556                117   
                        clearlow            12.183556                117   
                        clearlow            12.183556                117   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
                        overcast             6.609126                116   
       

In [60]:
df_ml.index

MultiIndex(levels=[['Greece', 'Switzerland'], ['P1EL', 'P2EL', 'P3EL', 'P4EL', 'P5EL', 'P6EL'], ['social', 'work'], ['clearhigh', 'clearlow', 'overcast']],
           labels=[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

### Preprocesing Phase

From the dataframe we can see that there are 5 catgorical features:
1. __*Country*__ 
2. __*Gender*__ 
3. __*Pattern*__ 
4. __*Context*__ 
5. __*SkyType*__ 

We can confirm this by checking the type of data for each feature:

In [28]:
x_fdata.dtypes

ID                                   int64
Country                             object
Gender                              object
Pattern                             object
Context                             object
SkyType                             object
contrast_mean_mSC5                 float64
contrast_max_mSC5                    int64
contrast_mean_Michelson            float64
contrast_mean_Michelson_cube123    float64
contrast_mean_Michelson_cube3      float64
contrast_mean_RMS                  float64
contrast_mean_RMS_cube123          float64
contrast_mean_RMS_cube3            float64
complexity_mean_JPEGtoBMP          float64
complexity_cube3_JPEGtoBMP         float64
dtype: object

As we can see the five features mentionned above are the only categorical features (type object).

Let's inspect these features closely:

In [29]:
categorical_features = ['Country','Gender','Pattern','Context', \
                        'SkyType']
for feat in categorical_features:
    print(feat + ':')
    print(x_fdata[feat].value_counts())

Country:
Greece         828
Switzerland    762
Name: Country, dtype: int64
Gender:
Male      816
Female    774
Name: Gender, dtype: int64
Pattern:
P4EL    265
P5EL    265
P1EL    265
P3EL    265
P2EL    265
P6EL    265
Name: Pattern, dtype: int64
Context:
social    798
work      792
Name: Context, dtype: int64
SkyType:
clearlow     552
clearhigh    522
overcast     516
Name: SkyType, dtype: int64


In order to perform dummy variable encoding on the categorical features, we can use the pandas method `pd.get_dummies()`. Also since we need k-1 dummy variables to represent k categories, we can drop the first column for each encoding (`drop_first = True`). We'll store this as a new dataframe `dummy_df`.

In [30]:
dummy_df = pd.get_dummies(x_fdata, columns=categorical_features, drop_first=False)
dummy_df.head(3)

ID  contrast_mean_mSC5  contrast_max_mSC5  contrast_mean_Michelson  \
0  11            5.023496                 82                 0.881714   
1  17            5.023496                 82                 0.881714   
2  29            5.023496                 82                 0.881714   

   contrast_mean_Michelson_cube123  contrast_mean_Michelson_cube3  \
0                         0.822095                       0.696552   
1                         0.822095                       0.696552   
2                         0.822095                       0.696552   

   contrast_mean_RMS  contrast_mean_RMS_cube123  contrast_mean_RMS_cube3  \
0         104.887032                 116.926766               138.288359   
1         104.887032                 116.926766               138.288359   
2         104.887032                 116.926766               138.288359   

   complexity_mean_JPEGtoBMP        ...         Pattern_P2EL  Pattern_P3EL  \
0                   0.101518        ...                    0             0   
1                   0.101518        ...                    0             0   
2                   0.101518        ...                    0             0   

   Pattern_P4EL  Pattern_P5EL  Pattern_P6EL  Context_social  Context_work  \
0             0             0             0               1             0   
1             0             0             0               1             0   
2             0             0             0               1             0   

   SkyType_clearhigh  SkyType_clearlow  SkyType_overcast  
0                  1                 0                 0  
1                  1                 0                 0  
2                  1                 0                 0  

[3 rows x 26 columns]

We standardize the features in preparation for the training (output excluded, "exciting" column).

In [31]:
x_without_std = dummy_df.copy() #x_without_std is the final standardized data (Just in case you want to use it)
for feat in dummy_df.columns.tolist():
    mean = dummy_df[feat].mean()
    std = dummy_df[feat].std()
    dummy_df[feat] = (dummy_df[feat] - mean)/std

In [32]:
x_data = dummy_df #x_data will be the final standardized data